### PR TITLE
state management and architecture decision

### PR DESCRIPTION
**Pull request by buenaflor**

### CODE REVIEW

File: doc/decisions/frontend_architecture_pattern.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# Frontend Architecture Pattern
+
+## Problem
+
+Permaplant needs to be designed with a consistent and maintainable architecture pattern.
```
Comment (markus2330) [inline doc/decisions/frontend_architecture_pattern.md:5]:
```suggestion
Permaplant needs to be designed with a documented, consistent and maintainable architecture pattern.
```

File: doc/decisions/frontend_architecture_pattern.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+## Constraints
+
+- The architecture pattern should be easy to understand and follow
```
Comment (markus2330) [inline doc/decisions/frontend_architecture_pattern.md:9]: No need to repeat it in every line:

```suggestion
The architecture pattern should be

- easy to understand and follow
```

File: doc/decisions/frontend_architecture_pattern.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+- The architecture pattern should be easy to test
+- The architecture pattern should be easy to scale
+- The architecture pattern should be suitable for a large, complex application
+- The architecture pattern should promote code reuse
+- The architecture pattern should be well-documented and widely used in the React community
+
+## Considered Alternatives
+
+- [Flux](https://reactjs.org/blog/2014/05/06/flux.html)
```
Comment (markus2330) [inline doc/decisions/frontend_architecture_pattern.md:18]: Please write a sentence why we won't use it.

File: doc/decisions/frontend_architecture_pattern.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+- [MVVM](https://www.detroitlabs.com/blog/intro-to-mvvm-in-react-with-mobx/)
+
+## Decision
+
+[Bulletproof React](https://github.com/alan2207/bulletproof-react) will be used as frontend architecture and guideline for PermaplanT.
```
Comment (markus2330) [inline doc/decisions/frontend_architecture_pattern.md:23]:
```suggestion
[Bulletproof React](https://github.com/alan2207/bulletproof-react) will be used as frontend architecture and guideline for PermaplanT.
Choices or derivations of this guidelines will be documented in `doc/architecture`.
```

File: doc/decisions/frontend_architecture_pattern.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+## Rationale
+
+Bulletproof React is a set of best practices and conventions that promote code reuse and make it easy to reason about the structure of the application. It's simple, easy to understand, and easy to test. This makes it a good choice for a large, complex application like Permaplant. Additionally, it is well-documented and widely used in the React community, which means that there is a lot of resources and support available for it.
```
Comment (markus2330) [inline doc/decisions/frontend_architecture_pattern.md:27]: Please use a new-line after every ".".

File: doc/decisions/frontend_architecture_pattern.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+## Implications
+
+## Related Decisions
+
+## Notes
+ - "bulletproof react" architecture pattern is developed by the team at Level Up Tutorials. It is not an official pattern by React.
```
Comment (markus2330) [inline doc/decisions/frontend_architecture_pattern.md:34]:
```suggestion

 - "bulletproof react" architecture pattern is developed by the team at Level Up Tutorials. It is not an official pattern by React.
```

File: doc/decisions/frontend_state_management.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+PermaplanT is a web application built with React that requires efficient and flexible state management.
+
+## Constraints
+
+- The state management library must support React and work seamlessly with the rest of the application's technology stack.
+- The library should be well-maintained, have a large community of developers, and a good documentation.
+
+## Considered Alternatives
+
+- [Redux](https://redux.js.org/)
```
Comment (markus2330) [inline doc/decisions/frontend_state_management.md:14]: Again, please a sentence for each alternative why it is not taken.
Which features do they have and why don't we need them?

File: doc/decisions/frontend_state_management.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+- [React context](https://reactjs.org/docs/context.html)
+- [MobX](https://mobx.js.org/)
+
+## Decision
+
+[Zustand](https://github.com/pmndrs/zustand) will be used as the state management library for PermaplanT.
+
+## Rationale
+
+Zustand is a lightweight and easy-to-use library that uses hooks, which makes it easy to integrate with React. Its simplicity also reduces the amount of boilerplate code and the need for complex setup and configuration.
```
Comment (markus2330) [inline doc/decisions/frontend_state_management.md:24]:
```suggestion
Zustand is a lightweight and easy-to-use library that uses hooks, which makes it easy to integrate with React.
Its simplicity also reduces the amount of boilerplate code and the need for complex setup and configuration.
```

File: doc/decisions/frontend_state_management.md [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+## Implications
+
+## Related Decisions
+ - Choosing React as the frontend library for PermaplanT
```
Comment (markus2330) [inline doc/decisions/frontend_state_management.md:29]: Please newlines after titles and please add link to the other decision.

**Review by markus2330 — CHANGES_REQUESTED** [review]
I tend to agree with the decisions but some explanations etc. is missing.

**Comment by markus2330** [comment]
Ideally you already add to doc/architecture the main choices&differences to Bulletproof React, otherwise it is only a half decision.

**Comment by Bushuo** [comment]
Hi, I heard great things about Zustand but never used it myself. I usually just went with Apollo Client as I mostly developed my personal React applications with a GraphQL backend. 

But I would like to point out, that I would not use Zustand in the way it is currently used inside the `seeds` feature.
React Query is a better tool for handling server state.

Just a quick pointer about state.
We can usually distinguish between three types of state inside a React application.
1. Local Component State -> is the dropdown open or not, is the checkbox clicked or not, and so on.
   I would use the default react `useState` for this or `Context` + `useReducer` for more complex local state

2. Application State -> Synchronous State. Local user preferences, UI state in a visual desgin app
   This is where 'Zustand' should be used

3. Server State -> Async state. A network request is needed before any state can be derived.
   I would use 'React Query' for this.

My rationale for this is, React Query has integrated mechanisms for network error states, deduping, retries, optimistic ui and so on. Thus much of the wiring code that would be needed in the local state management approach is just included.

Here is the documentation that points out the differences to local state management tools
https://react-query-v3.tanstack.com/guides/does-this-replace-client-state
And a quick 100 second summary of React Query
https://www.youtube.com/watch?v=novnyCaa7To

**Comment by buenaflor** [comment]
@Bushuo Yup, makes sense and I agree with this.

File: doc/decisions/frontend_state_management.md [hunk 10]
```diff
@@ -0,0 +1,41 @@
+# Frontend State Management Library
+
+## Problem
+
+PermaplanT is a web application built with React that requires efficient and flexible state management.
```
Comment (Bushuo) [inline doc/decisions/frontend_state_management.md:5]:
```suggestion
PermaplanT is a web application built with React that requires efficient and flexible state management.
Inside a React application we can usually distinguish between three types of state.

- Local Component State: Is the dropdown open or not, is the link active or not, and so on.

- Global Application State (Synchronous): Local user preferences, Sidenav is open, UI state in a visual design app.

- Server State (Asynchronous): A network request is needed before any state can be derived.
```

File: doc/decisions/frontend_state_management.md [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+## Constraints
+
+- The state management library must support React and work seamlessly with the rest of the application's technology stack.
+- The library should be well-maintained, have a large community of developers, and a good documentation.
+
+## Considered Alternatives
+
```
Comment (Bushuo) [inline doc/decisions/frontend_state_management.md:13]: Let us please consider separately, global application state and server state.

File: doc/decisions/frontend_architecture_pattern.md (hunk 3 above)
Comment (absurd-turtle) [reply doc/decisions/frontend_architecture_pattern.md:20]: @buenaflor Will you look into this?

**Comment by absurd-turtle** [comment]
> 

Yes, it makes perfect sense to distinguish between the "types" of state.
I separated the concerns in the decision document.
Do you have any insights on why to use React Query specifically as opposed to vercel/swr for example?

**Comment by Bushuo** [comment]
> @Bushuo
> 
> Yes, it makes perfect sense to distinguish between the "types" of state. I separated the concerns in the decision document. Do you have any insights on why to use React Query specifically as opposed to vercel/swr for example?

Yup, swr is more minimalistic for easy use cases. React Query also has for example, builtin features for pagination. I think in our case a more holistic solution should be preferred.

File: doc/decisions/frontend_architecture_pattern.md (hunk 3 above)
Comment (Bushuo) [reply doc/decisions/frontend_architecture_pattern.md:20]: @absurd-turtle 
May I give an answer to that.
Flux is what was used before Redux got invented. It is pretty similar. Redux implements the Flux concepts in a library with a little more constraints.
The same reasons apply here, why we should not use Flux.

File: doc/decisions/frontend_architecture_pattern.md [hunk 12]
```diff
@@ -0,0 +1,44 @@
+# Frontend Architecture Pattern
+
+## Problem
+
+Permaplant needs to be designed with a documented, consistent and maintainable architecture pattern.
```
Comment (Bushuo) [inline doc/decisions/frontend_architecture_pattern.md:5]:
```suggestion
PermaplanT needs to be designed with a documented, consistent and maintainable architecture pattern.
```

**Review by Bushuo — APPROVED** [review]

File: doc/decisions/frontend_architecture_pattern.md [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+## Decision
+
+[Bulletproof React](https://github.com/alan2207/bulletproof-react) will be used as frontend architecture and guideline for PermaplanT.
+Choices or derivations of this guidelines will be documented in `doc/architecture`.
+
+## Rationale
+
+Bulletproof React is a set of best practices and conventions that promote code reuse and make it easy to reason about the structure of the application.
+It's simple, easy to understand, and easy to test.
+This makes it a good choice for a large, complex application like Permaplant.
```
Comment (Bushuo) [inline doc/decisions/frontend_architecture_pattern.md:35]:
```suggestion
This makes it a good choice for a large, complex application like PermaplanT.
```

File: doc/decisions/frontend_architecture_pattern.md (hunk 3 above)
Comment (markus2330) [reply doc/decisions/frontend_architecture_pattern.md:20]: Please simply push such statements into the PR. Later readers will not read what we wrote here but only what is in the decision text.

**Comment by markus2330** [comment]
@Bushuo please also add all the details you know about the different state libs. Feel free to push here.

**Comment by Bushuo** [comment]
Hi @markus2330
> Please also add all the details you know about the different state libs. Feel free to push here.

The ones I know of are either already included (Redux), or related to GraphQL.
I think the decision is ready to be merged.

File: doc/decisions/frontend_architecture_pattern.md [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+PermaplanT needs to be designed with a documented, consistent and maintainable architecture pattern.
+
+## Constraints
+
+The architecture pattern should be
+
+- easy to understand and follow
+- be easy to test
+- be easy to scale
+- be suitable for a large, complex application
```
Comment (markus2330) [inline doc/decisions/frontend_architecture_pattern.md:14]:
```suggestion
- be suitable for a large, highly-interactive application
```

File: doc/decisions/frontend_architecture_pattern.md [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+This makes it a good choice for a large, complex application like PermaplanT.
+Additionally, it is well-documented and widely used in the React community, which means that there is a lot of resources and support available for it.
+
+## Implications
+
+## Related Decisions
+
+## Notes
+
+ - "bulletproof react" architecture pattern is developed by the team at Level Up Tutorials. It is not an official pattern by React.
```
Comment (markus2330) [inline doc/decisions/frontend_architecture_pattern.md:44]:
```suggestion
 - "bulletproof react" architecture pattern is developed by the team at Level Up Tutorials.
 It is not an official pattern by React.
```

**Review by markus2330 — APPROVED** [review]
Great work, good that we looked into this again!

**Pull request merged**